Simplify generated build

jamesfredley opened the pull request:
This was significantly more involved than anticipated.

This PR:

- Addresses the 5 linked issues
- Consolidates Gradle Plugin configuration to `buildscript{}`, `plugins{}` and `apply plugin:` in `build.gradle`
- Changes Gradle buildSrc/build.gradle and Settings File to non-default, optional features
- When the Gradle Settings File feature is selected, it only contains one line defining the root project
- When the Gradle buildSrc feature is selected `buildscript{}` is not generated, they are equivalent
- Works around https://github.com/grails/grails-gradle-plugin/issues/351 by using "apply plugin:" for Gradle plugin ids that are not published to the Gradle Plugin repo and are not release versions.  Release versions work fine with `id` and `version`.
- significant updates to fix the tests, including one that was not cross platform compatible
- adds classpath scope
- updates asset-pipeline to 4.5.1
- standardizes on double quotes with no parentheses in build.gradle, did not update web-driver/selenium since it has been removed in 7.0.x
- dependencies set to pom will still have parentheses, since they are required for platform()
- move hamcrest, micronaut-http-client and neo4j-harness dependencies to features instead of manually being set in dependencies.rocker.raw
- remove dependencies.rocker.raw since it duplicates toSnippet()
- rename RUNTIME to RUNTIME_ONLY, TEST to TEST_IMPLEMENTATION and
TEST_RUNTIME to TEST_RUNTIME_ONLY to reflect the actual Gradle scopes
- rename compile to implementation

Merging up to 7.0.x is going to be fun.

jamesfredley committed:
Update asset-pipeline to 4.5.1

jamesfredley committed:
remove parenthesis from Gradle dependencies

jamesfredley committed:
consolidate Gradle plugin config, make Gradle buildSrc and settings file optional features

when buildSrc feature is selected buildscript{} will not generate in main build.gradle file

jamesfredley committed:
fix failing tests after Gradle plugin configuration consolidation

jamesfredley committed:
fix failing test after Gradle plugin configuration consolidation

jamesfredley commented:
This required significant changes to get the desired outcome. We need to debate whether this is completed in 6.2.2 or 7.0.0.

https://github.com/grails/gorm-docs/issues/24 - ticket to update documentation.

jamesfredley commented on grails-forge-core/src/main/java/org/grails/forge/feature/assetPipeline/AssetPipeline.java:
```diff
@@ -60,13 +60,11 @@ public String getDescription() {
 
     @Override
     public void apply(GeneratorContext generatorContext) {
-        generatorContext.addBuildscriptDependency(Dependency.builder()
```
Only the plugin id is required.

matrei commented on grails-forge-core/src/main/resources/pom.xml:
```diff
@@ -125,12 +125,12 @@
             <groupId>com.bertramlabs.plugins</groupId>
             <artifactId>asset-pipeline-grails</artifactId>
-            <version>4.3.0</version>
+            <version>4.5.1</version>
```
There has been a recent release `4.5.2` of `asset-pipeline` with some relevant changes for Grails.

jamesfredley replied:
I missed that, since the tag was not pushed to github and I didn't double check maven central.  I'll update.

matrei commented on grails-forge-core/src/main/resources/pom.xml:
```diff
@@ -125,12 +125,12 @@
             <artifactId>asset-pipeline-gradle</artifactId>
-            <version>4.3.0</version>
+            <version>4.5.1</version>
         </dependency>
```
There has been a recent release `4.5.2` of `asset-pipeline` with some relevant changes for Grails.

matrei replied:
`4.5.2` will default to `skipNonDigests = true` which means that it can be removed from the `assets` block in `build.gradle`.
`minifyJs` and `minifyCss` are already defaulting to `true`, so that means the entire `assets` block can be removed (if we don't want to explicitly show it in the build file).

jamesfredley replied:
Updated to 4.5.2

I think the removal of assets {} should wait for 7.0.0 and the outcome of https://github.com/orgs/grails/discussions/13829

jamesfredley replied:
4.5.2 has not been released to https://plugins.gradle.org/plugin/com.bertramlabs.asset-pipeline, so going back to 4.5.1 for this PR

`Plugin [id: 'com.bertramlabs.asset-pipeline', version: '4.5.2'] was not found in any of the following sources:`

jamesfredley replied:
https://github.com/bertramdev/asset-pipeline/issues/358

matrei replied:
Hmm, it is available here https://mvnrepository.com/artifact/com.bertramlabs.plugins/asset-pipeline-gradle

I wonder if something went wrong or if it takes longer to show up in the Gradle Plugin Portal.
@davydotcom Do you know, why it's not available yet?

matrei commented on grails-forge-core/src/main/java/org/grails/forge/feature/build/gradle/templates/dependency.rocker.raw:
```diff
@@ -1,3 +1,3 @@
 @args(String groupId, String artifactId, String scope, String version, boolean pom)
 
-@(pom ? scope + " platform" : scope)("@groupId:@artifactId@(version != null ? ":" + version : "")")
+@(pom ? scope + " platform" : scope) "@groupId:@artifactId@(version != null ? ":" + version : "")"
```
Will this add the required parenthesis for `platform`?

matrei replied:
I tried to add the `grails-bom` to the `base` feature and the parenthesis did come out ok.
However, there was a spacing issues: `implementation platform ( "org.grails:grails-bom:6.2.2")`

jamesfredley replied:
The extra spaces were coming from https://github.com/grails/grails-forge/pull/424/files#diff-ae814cab294ecd3c3a3fdfd4f7df62fbdaa5653fac1f6f0451f267265c7395ee, which is now fixed.

But, I found that the above code, which is doing the same thing, is only used for 3 dependencies where the Gradle scope has not been implemented in the rest of grails-forge.  I am going to add the missing scopes (`testImplementation` and `testRuntimeOnly`), move those over to remove dependency.rocker.raw, since toSnippet() is all we will need. 
https://github.com/grails/grails-forge/blob/6.2.x/grails-forge-core/src/main/java/org/grails/forge/feature/build/gradle/templates/dependencies.rocker.raw#L20

jamesfredley replied:
https://github.com/grails/grails-forge/pull/424/commits/3647ba8eb162a7ec692a59d84337f47a503e070d fully implements `testImplementation` and `testRuntimeOnly` which were there but needed clearer naming.

matrei commented on grails-forge-core/src/test/groovy/org/grails/forge/feature/lang/GrailsApplicationSpec.groovy:
```diff
@@ -28,7 +28,7 @@ class GrailsApplicationSpec extends BeanContextSpec implements CommandOutputFixt
 
         then:
-        buildGradle.contains('mainClass.set("example.grails.Application")')
+        buildGradle.contains('mainClass.set("example.grails.Application"')
```
This test passes, but the removal of the trailing parenthesis is not warranted.

matrei replied:
Actually, this is java/kotlin format, the groovy syntax is nicer:
```groovy
mainClass = 'example.grails.Application'
```

jamesfredley replied:
Updated the rocker template and the test.

matrei committed:
fix: Tests working on Windows

matrei committed:
test: Add tests for optional build file features

matrei reviewed (commented):
This was some heavy lifting, nice work James!

I think there is a risk that doing this big of a change will introduce bugs in `6.2.2`. As you say, we should discuss if it is worth the trouble to merge this to `6.2.x`.

I'm not sure it's even going to help anyone out, but rather cause more confusion. If a user is on Grails 5 or lower, is there any need to upgrade to Grails 6 before upgrading to Grails 7? If a user is on Grails 6 it should be an easy operation to remove buildSrc when upgrading to Grails 7 (if that is what the user wants).

jamesfredley committed:
Update asset-pipeline to 4.5.2

jamesfredley committed:
Merge branch 'simplify-generated-build' of https://github.com/grails/grails-forge into simplify-generated-build

jamesfredley committed:
Update asset-pipeline to 4.5.2 in test

jamesfredley committed:
Adjust mainClass syntax

codeconsole commented:
> I'm not sure it's even going to help anyone out, but rather cause more confusion. If a user is on Grails 5 or lower, is there any need to upgrade to Grails 6 before upgrading to Grails 7? If a user is on Grails 6 it should be an easy operation to remove buildSrc when upgrading to Grails 7 (if that is what the user wants).

@matrei we need to think from the perspective of what is `create-app` for?  Creating new apps. 

1. Why create an overcomplicated build for a new app only for someone to have to dramatically change it in an upcoming release?
2. Grails <= 6.2.0 was broken.  We were just cleaning up an inherited mess and are putting our names behind the release.  Let's finish with a more polished product we can be proud o.
3. This has to be done either way for Grails 7.
4. This can continue to evolve independently as it is only forge. 
5. The current behavior is inconsistent with Grails shell.  Does Grails shell create-app create a buildSrc?

jamesfredley committed:
fix spacing when using bom/platform

jamesfredley committed:
back to asset-pipeline 4.5.1

jamesfredley committed:
move hamcrest, micronaut-http-client and neo4j-harness dependencies to features
instead of manually being set in dependencies.rocker.raw
remove dependencies.rocker.raw since it duplicates toSnippet()

rename RUNTIME to RUNTIME_ONLY, TEST to TEST_IMPLEMENTATION and
TEST_RUNTIME to TEST_RUNTIME_ONLY to reflect the actual Gradle scope

matrei commented on grails-forge-core/src/main/java/org/grails/forge/build/dependencies/Scope.java:
```diff
@@ -30,13 +30,14 @@ public class Scope {
     public static final Scope TEST_ANNOTATION_PROCESSOR = new Scope(Source.TEST, Collections.singletonList(Phase.ANNOTATION_PROCESSING));
     public static final Scope TEST_COMPILE_ONLY = new Scope(Source.TEST, Collections.singletonList(Phase.COMPILATION));
-    public static final Scope TEST_RUNTIME = new Scope(Source.TEST, Collections.singletonList(Phase.RUNTIME));
+    public static final Scope TEST_RUNTIME_ONLY = new Scope(Source.TEST, Collections.singletonList(Phase.RUNTIME));
```
Should we also add the `integrationTest*` scopes for completeness. Geb, Selenium and perhaps some others should preferably be moved to those.

And what about renaming `compile` to `implementation`?

jamesfredley replied:
Good catch, I've renamed `compile` to `implementation`.

For `integrationTest*` scopes, I'd like to do that on the `7.0.x` branch since it already has INTEGRATION_TEST_IMPLEMENTATION_TEST_FIXTURES https://github.com/grails/grails-forge/blob/a875736a7a41f108ad5a714e70acf2bd5688d3e9/grails-forge-core/src/main/java/org/grails/forge/build/dependencies/Scope.java#L40

jamesfredley commented:
I've seen this test fail several times.  Deleting build manually gets it working again.

```
> Task :grails-forge-core:test FAILED

GradleSpec > test buildSrc/build.gradle FAILED
    Condition not satisfied:

    buildSrcGradle
    |
    ""
        at org.grails.forge.build.gradle.GradleSpec.test buildSrc/build.gradle(GradleSpec.groovy:50)
```

jamesfredley committed:
rename compile to implementation

jamesfredley committed:
add grails-bom to buildSrc/build.gradle to set dependency versions the same as main build.gradle file

this is the only Gradle build dependency that is needed in buildSrc/build.gradle and not in buildscript{}
